spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaBinderMetrics.java

```diff
@@ -0,0 +1,137 @@
+
+	private final BindingServiceProperties bindingServiceProperties;
+
+	private final PropertyResolver propertyResolver;
+
+	private ConsumerFactory<?, ?> defaultConsumerFactory;
+
+	public KafkaBinderMetrics(KafkaMessageChannelBinder binder, KafkaBinderConfigurationProperties binderConfigurationProperties,
+			BindingServiceProperties bindingServiceProperties, PropertyResolver propertyResolver, ConsumerFactory<?, ?> defaultConsumerFactory) {
+		this.binder = binder;
+		this.binderConfigurationProperties = binderConfigurationProperties;
+		this.bindingServiceProperties = bindingServiceProperties;
+		this.propertyResolver = propertyResolver;
+		this.defaultConsumerFactory = defaultConsumerFactory;
+	}
+
+	public KafkaBinderMetrics(KafkaMessageChannelBinder binder, KafkaBinderConfigurationProperties binderConfigurationProperties,
+			BindingServiceProperties bindingServiceProperties, PropertyResolver propertyResolver) {
+		this(binder, binderConfigurationProperties, bindingServiceProperties, propertyResolver, null);
+	}
+
+	@Override public Collection<Metric<?>> metrics() {
+		List<Metric<?>> metrics = new LinkedList<>();
+		for (Map.Entry<String, KafkaMessageChannelBinder.TopicInformation> topicInfo : this.binder.getTopicsInUse().entrySet()) {
+			if (!topicInfo.getValue().isConsumerTopic()) {
+				continue;
+			}
+
+			String topic = topicInfo.getKey();
+			String binding = topic;
+			for(Map.Entry<String, BindingProperties> bindingProperties : bindingServiceProperties.getBindings().entrySet()) {
+				if(bindingProperties.getValue().getDestination() != null && bindingProperties.getValue().getDestination().equals(topic)) {
+					binding = bindingProperties.getKey();
+				}
+			}
+			String group = propertyResolver.getProperty("spring.cloud.stream.bindings." + binding + ".group");
+			if(group == null) {
+				continue;
+			}
+
+			try (Consumer<?, ?> metadataConsumer = createConsumerFactory(group).createConsumer()) {
+				List<PartitionInfo> partitionInfos = metadataConsumer.partitionsFor(topic);
+				List<TopicPartition> topicPartitions = new LinkedList<>();
+				for (PartitionInfo partitionInfo : partitionInfos) {
+					topicPartitions.add(new TopicPartition(partitionInfo.topic(), partitionInfo.partition()));
+				}
+				Map<TopicPartition, Long> endOffsets = metadataConsumer.endOffsets(topicPartitions);
+				long lag = 0;
+				for (Map.Entry<TopicPartition, Long> endOffset : endOffsets.entrySet()) {
+					OffsetAndMetadata current = metadataConsumer.committed(endOffset.getKey());
+					if (current != null) {
+						lag += endOffset.getValue() - current.offset();
+					} else {
+						lag += endOffset.getValue();
+					}
+				}
+				metrics.add(new Metric<>(String.format("%s.%s.%s.lag", METRIC_PREFIX, binding, group), lag));
+			} catch (Exception e) {
+				LOG.debug("Cannot generate metric for topic: " + topic, e);
+			}
+		}
+		return metrics;
+	}
+
+	private ConsumerFactory<?,?> createConsumerFactory(String group) {
```

Comment:
I was also thinking about this, but preferred to keep the code simpler now to make it easier to review and merge. Can we postpone this to the moment after we merge this PR? I can handle this, but I prefer little steps approach. But it is up to you to decide :) .